src/ctl.c

```diff
@@ -3083,3 +3085,57 @@ experimental_hooks_remove_ctl(tsd_t *tsd, const size_t *mib, size_t miblen,
 label_return:
 	return ret;
 }
+
+static int
+experimental_frag_ctl(tsd_t *tsd, const size_t *mib, size_t miblen,
+    void *oldp, size_t *oldlenp, void *newp, size_t newlen) {
+	int ret = 0;
+
+	/* Requires both read and write */
+	if (oldp == NULL || oldlenp == NULL || newp == NULL || newlen == 0) {
+		ret = EINVAL;
+		goto label_return;
+	}
+
+	/*
+	 * Make sure newlen is an exact multiple of sizeof(const void *),
+	 * and if so, reduce it to the multiple i.e. count of input pointers.
+	 * Rely on compiler to convert multiplication / division to shifts.
+	 */
+	if (newlen != newlen / sizeof(const void *) * sizeof(const void *)) {
+		ret = EINVAL;
+		goto label_return;
+	}
+	newlen /= sizeof(const void *);
+
+	/*
+	 * Make sure that *oldlenp is an exact multiple of sizeof(size_t) * 3,
+	 * and that the multiple equals the count of input pointers.
+	 */
+	if (*oldlenp != newlen * sizeof(size_t) * 3) {
+		ret = EINVAL;
+		goto label_return;
+	}
+
+	/*
+	 * Each pointer will output three entries for the extent it resides in:
+	 * (a) the number of free regions,
+	 * (b) the total number of regions, and
+	 * (c) the total size of the extent in terms of bytes.
+	 */
+	void * const newend = (const void **)newp + newlen;
+	for (; newp < newend; newp = (const void **)newp + 1) {
+		const extent_t *ext = iealloc(
+		    tsd_tsdn(tsd), *(const void **)newp);
+		if (unlikely(ext == NULL)) {
+			memset(oldp, 0, sizeof(size_t) * 3);
```

Comment:
Not really need to zero it out. As long as we return an error code, the user is not supposed to look into the values anyway.